### PR TITLE
Implemented most tokens

### PR DESCRIPTION
**Pull request by wadealexc**
Does not handle whitespace, some string literals, or identifiers. Added tests for all implemented tokens.

### CODE REVIEW

**Comment by wadealexc** [comment]
Lexer refactor is 90% done. Immediate next steps:
1. Fix the failing test in lex_tests.rs
2. Add more tests to lex_tests
3. Clean TODOs and stray comments

Then:
4. Move the new helper functions I wrote for lex_4_25::test to a module that other tests can access
5. Start work on parser

Longer term:
6. There's still some redundancy and inefficiency that can be improved upon
7. A few unimplemented tokens (>>> and >>>=)

**Comment by wadealexc** [comment]
Okay -- lexer now recognizes almost everything!

Added an edge case to identifier matching to handle hex literals (hex"DEADBEEF"). 2 cases I know I don't handle: escape characters in strings, and leading '.' in rational numbers.

File: tests/parse_tests.rs [hunk 1]
```diff
@@ -33,7 +33,8 @@ mod parse_tests {
                     node: lex_4_25::Token::Pragma,
                     children: vec![
                         Box::new(lex_4_25::Token::Identifier("solidity".to_string()).to_leaf()),
-                        Box::new(lex_4_25::Token::Version("^0.4.25".to_string()).to_leaf())
+                        Box::new(lex_4_25::Token::BitwiseXor.to_leaf()),
```
Comment (jalextowle) [inline tests/parse_tests.rs:36]: I think we'll probably want to make the AST for this look like:

```
Pragma
    | Identifier("solidity")
    | Version
        | BitwiseXor
        | VersionNumber("0.4.25")
```

This looks good for now though. I'll add that to the AST issue

File: tests/parser_integration_tests.rs [hunk 2]
```diff
@@ -17,7 +17,8 @@ mod parser_integration_tests {
                     node: lex_4_25::Token::Pragma,
                     children: vec![
                         Box::new(lex_4_25::Token::Identifier("solidity".to_string()).to_leaf()),
-                        Box::new(lex_4_25::Token::Version("^0.4.25".to_string()).to_leaf())
+                        Box::new(lex_4_25::Token::BitwiseXor.to_leaf()),
```
Comment (jalextowle) [inline tests/parser_integration_tests.rs:20]: Same as other AST comment.

File: src/lex_4_25.rs [hunk 3]
```diff
@@ -53,6 +54,8 @@ pub enum Token {
     CloseParenthesis,
     Colon,
     Comma,
+    CommentMulti,
```
Comment (jalextowle) [inline src/lex_4_25.rs:57]: I'm glad you added these. We need to start dealing with comments, and I think that the lexer is probably the right place for them.

File: src/lex_4_25.rs [hunk 4]
```diff
@@ -596,161 +692,481 @@ fn match_collected(collected: String) -> Token {
...
 
 /**
- * Returns the next Token found in the line and increments cur
- * to the end of the Token in the parsed line
+ * Matches . at line[*cur] with Token::Dot
  */
-pub fn next_token(line: &Vec<char>, cur: &mut usize) -> Token {
-    let mut string = false;
-    let mut collected = String::new();
+fn match_period(line: &Vec<char>, cur: &mut usize) -> Token {
```
Comment (jalextowle) [inline src/lex_4_25.rs:697]: What is the benefit of using this function over just using `Token::Dot` in the `match` statement?

On a side note, does it make sense to call periods "dots" in the Token enum? (Depending on your thoughts we can open an issue to change that. I'm not stuck on it)

File: src/lex_4_25.rs [hunk 5]
```diff
@@ -765,40 +1181,971 @@ pub fn peek_token(line: &Vec<char>, cur: &mut usize) -> Token {
...
-        let actual = next_token(&chars, cur);
-        match actual {
-            Token::StringLiteral(path) => assert_eq!(&path, "\"\""), 
-            actual => panic!("Expected: {:?} | Actual: {:?}", Token::StringLiteral("\"\"".to_string()), actual)
-        }
+    fn test_pragma1() {
+        let s = to_chars("^0.4.25;");
+        let cur = &mut 0;
+        expect_next_token(&s, cur, Token::BitwiseXor);
+        expect_next_token(&s, cur, Token::Version(String::from("0.4.25")));
```
Comment (jalextowle) [inline src/lex_4_25.rs:1206]: I think we should check for the semicolon token after the version

File: tests/lex_tests.rs [hunk 6]
```diff
@@ -4,79 +4,66 @@ extern crate solfix;
 mod lexer_tests {
     use solfix::lex_4_25;
 
+    fn fail_test(expect: lex_4_25::Token, actual: lex_4_25::Token) {
+        panic!("Expected: {:?} | Actual: {:?}", expect, actual);
+    }
+
+    fn expect_next_token(s: &Vec<char>, cur: &mut usize, t: lex_4_25::Token) {
```
Comment (jalextowle) [inline tests/lex_tests.rs:11]: I'm a huge fan of this function! (No changes needed to resolve)

File: src/lex_4_25.rs [hunk 7]
```diff
@@ -596,161 +692,481 @@ fn match_collected(collected: String) -> Token {
...
+                decimal_found = true;
+            }
+        } else if line.match_idx(*cur, 'e') || line.match_idx(*cur, 'E') {
+            // Cannot have 2 exponents, or an exponent in a version
+            if exponent_found || version_found {
+                return Token::Illegal;
             } else {
-                collected.push(line[*cur]);
+                // If we find an exponent, there must be at least 1 more digit
+                // (Trailing decimals are allowed, though!)
```
Comment (jalextowle) [inline src/lex_4_25.rs:1084]: Weird! (No changes needed to resolve)

File: src/lex_4_25.rs [hunk 8]
```diff
@@ -596,161 +692,481 @@ fn match_collected(collected: String) -> Token {
...
-        other => other
+    *cur -= 1;
+    Token::DecimalNumber(collected)
+}
+
+/**
+ * Returns the next Token found in the line and increments cur
+ * to the end of the Token in the parsed line
+ */
+pub fn next_token(line: &Vec<char>, cur: &mut usize) -> Token {
```
Comment (jalextowle) [inline src/lex_4_25.rs:1111]: This is much cleaner than before. This function is very nicely done; great job on this one! (No changes needed to resolve)

File: src/lex_4_25.rs [hunk 9]
```diff
@@ -379,40 +383,132 @@ impl Token {
...
+    fn is_digit_at(&self, idx: usize) -> bool;
+    fn is_rational_at(&self, idx: usize) -> bool;
+    fn is_hex_digit_at(&self, idx: usize) -> bool;
+    fn is_hex_delim_at(&self, idx: usize) -> bool;
+    fn is_whitespace_at(&self, idx: usize) -> bool;
+}
+
+impl LineMatch for Vec<char> {
+    fn match_idx(&self, idx: usize, val: char) -> bool {
+        match self.get(idx) {
```
Comment (jalextowle) [inline src/lex_4_25.rs:397]: `Vec::get` is really cool. I'd like to start using this pattern in the parser if possible, where all of the errors bubble up using the Option or Result types. (No changes needed to resolve

**Review by jalextowle — CHANGES_REQUESTED** [review]
I have a couple of nits (match_period and the test comment), and after those are resolved this looks good to go. 

Overall very nice job on this!

File: src/lex_4_25.rs (hunk 4 above)
Comment (wadealexc) [reply src/lex_4_25.rs:697]: Technically `var x = .14;` is allowed, so a leading period could be a rational or part of an identifier, depending on what comes next. I was trying to figure out how to best handle this case and then I remembered that leading digits aren't allowed in identifiers. So I'll just check the next character and if it's a digit, read a rational.

I'm fine with `Dot`. `Period` is fine too. Not a big deal either way.

**Comment by wadealexc** [comment]
Last commit addresses requested changes!

File: src/lex_4_25.rs (hunk 4 above)
Comment (jalextowle) [reply src/lex_4_25.rs:697]: Okay that makes sense. Okay, well if you're fine with `Dot`, I'd say that we should keep it. It's sort of an easter egg I guess

**Review by jalextowle — APPROVED** [review]
Signed off by: Alex Towle
